### PR TITLE
ARTEMIS-214 exception when sending 10k message

### PR DESCRIPTION
**Pull request by gaohoward**

### CODE REVIEW

File: artemis-protocols/artemis-proton-plug/src/main/java/org/proton/plug/util/DeliveryUtil.java [hunk 1]
```diff
@@ -30,6 +30,9 @@ public static int readDelivery(Receiver receiver, ByteBuf buffer) {
       while ((count = receiver.recv(buffer.array(), buffer.arrayOffset() + buffer.writerIndex(), buffer.writableBytes())) > 0) {
          // Increment the writer index by the number of bytes written into it while calling recv.
          buffer.writerIndex(buffer.writerIndex() + count);
+         if (buffer.writableBytes() <= 0) {
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-proton-plug/src/main/java/org/proton/plug/util/DeliveryUtil.java]: shouldn't this be if (buffer.writableBytes() < count)

you may get into a situation that you have space for just part of the buffer coming.

File: artemis-protocols/artemis-proton-plug/src/main/java/org/proton/plug/util/DeliveryUtil.java [hunk 2]
```diff
@@ -30,6 +30,9 @@ public static int readDelivery(Receiver receiver, ByteBuf buffer) {
...
+            buffer.ensureWritable(count);
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-proton-plug/src/main/java/org/proton/plug/util/DeliveryUtil.java]: Or maybe we don't need the if at all.. just always cal buffer.ensureWritable ?

Comment (gaohoward) [inline artemis-protocols/artemis-proton-plug/src/main/java/org/proton/plug/util/DeliveryUtil.java]: You are right Clebert, we don't need it. The netty api says if there are enough writeable bytes this call won't do anything. I'll remove the 'if' checking. Thanks

**Comment by clebertsuconic** [comment]
Please remove the word (ENTMQ-1558)...

That JIRA is probably your internal JIRA. There's no need to mention it here upstream. You can link ARTEMIS-214 at your internal JIRA, but I don't see a need to mention it here.

**Comment by gaohoward** [comment]
ok done.

**Comment by clebertsuconic** [comment]
@gaohoward the test you added doesn't work:

I get lots of exceptions:

```

java.lang.IllegalArgumentException: Unknown Destination Type passed to JMS Transformer.
	at org.apache.activemq.transport.amqp.message.JMSMappingOutboundTransformer.destinationType(JMSMappingOutboundTransformer.java:322)
	at org.apache.activemq.transport.amqp.message.JMSMappingOutboundTransformer.convert(JMSMappingOutboundTransformer.java:193)
	at org.apache.activemq.artemis.core.protocol.proton.converter.ProtonMessageConverter.outbound(ProtonMessageConverter.java:105)
	at org.apache.activemq.artemis.core.protocol.proton.plug.ProtonSessionIntegrationCallback.encodeMessage(ProtonSessionIntegrationCallback.java:214)
	at org.proton.plug.context.server.ProtonServerSenderContext.deliverMessage(ProtonServerSenderContext.java:270)
	at org.apache.activemq.artemis.core.protocol.proton.plug.ProtonSessionIntegrationCallback.sendMessage(ProtonSessionIntegrationCallback.java:363)
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.deliverStandardMessage(ServerConsumerImpl.java:1027)
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.proceedDeliver(ServerConsumerImpl.java:417)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.proceedDeliver(QueueImpl.java:2463)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.deliver(QueueImpl.java:1955)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.access$1400(QueueImpl.java:98)
	at org.apache.activemq.artemis.core.server.impl.QueueImpl$DeliverRunner.run(QueueImpl.java:2694)
	at org.apache.activemq.artemis.utils.OrderedExecutorFactory$OrderedExecutor$ExecutorTask.run(OrderedExecutorFactory.java:100)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
22:43:02,734 WARN  [org.apache.activemq.artemis.core.server] AMQ222151: removing consumer which did not handle a message, consumer=ServerConsumerImpl [id=0, filter=null, binding=LocalQueueBinding [address=jms.queue.exampleQueue, queue=QueueImpl[name=jms.queue.exampleQueue, postOffice=PostOfficeImpl [server=ActiveMQServerImpl::serverUUID=971a280e-0b58-11e6-999f-426c8fd39000]]@35137fbc, filter=null, name=jms.queue.exampleQueue, clusterName=jms.queue.exampleQueue971a280e-0b58-11e6-999f-426c8fd39000]], message=Reference[5]:RELIABLE:ServerMessage[messageID=5,durable=true,userID=null,priority=4, bodySize=32768, timestamp=Mon Apr 25 22:43:02 EDT 2016,expiration=0, durable=true, address=jms.queue.exampleQueue,properties=TypedProperties[JMS_AMQP_NATIVE=false,JMS_AMQP_MA_x-opt-jms-dest=0,JMS_AMQP_MESSAGE_FORMAT=0,JMS_AMQP_MA_x-opt-jms-msg-type=5]]@74198727: java.lang.IllegalStateException: Can't deliver message org.proton.plug.exceptions.ActiveMQAMQPInternalErrorException: Unknown Destination Type passed to JMS Transformer.
	at org.apache.activemq.artemis.core.protocol.proton.plug.ProtonSessionIntegrationCallback.sendMessage(ProtonSessionIntegrationCallback.java:370) [:]
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.deliverStandardMessage(ServerConsumerImpl.java:1027) [:]
	at org.apache.activemq.artemis.core.server.impl.ServerConsumerImpl.proceedDeliver(ServerConsumerImpl.java:417) [:]
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.proceedDeliver(QueueImpl.java:2463) [:]
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.deliver(QueueImpl.java:1955) [:]
	at org.apache.activemq.artemis.core.server.impl.QueueImpl.access$1400(QueueImpl.java:98) [:]
	at org.apache.activemq.artemis.core.server.impl.QueueImpl$DeliverRunner.run(QueueImpl.java:2694) [:]
	at org.apache.activemq.artemis.utils.OrderedExecutorFactory$OrderedExecutor$ExecutorTask.run(OrderedExecutorFactory.java:100) [:]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [rt.jar:1.8.0_73]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [rt.jar:1.8.0_73]
	at java.lang.Thread.run(Thread.java:745) [rt.jar:1.8.0_73]
Caused by: org.proton.plug.exceptions.ActiveMQAMQPInternalErrorException: Unknown Destination Type passed to JMS Transformer.
	at org.proton.plug.context.server.ProtonServerSenderContext.deliverMessage(ProtonServerSenderContext.java:274) [:]
	at org.apache.activemq.artemis.core.protocol.proton.plug.ProtonSessionIntegrationCallback.sendMessage(ProtonSessionIntegrationCallback.java:363) [:]
	... 10 more

```

**Comment by gaohoward** [comment]
sorry about that. I will investigate.

**Comment by gaohoward** [comment]
I belive this is caused by the recent replacement of proton-jms with activemq-amqp lib.
In activemq-amqp, it checks the message's destination like that:

   private static byte destinationType(Destination destination) {
      if(destination instanceof Queue) {
         return (byte)(destination instanceof TemporaryQueue?2:0);
      } else if(destination instanceof Topic) {
         return (byte)(destination instanceof TemporaryTopic?3:1);
      } else {
         throw new IllegalArgumentException("Unknown Destination Type passed to JMS Transformer.");
      }
   }
Unfortunately the destination we have is ServerDestination which is only an instance of Destination.
We probably need to subclasses that implements Queu and Topic jms interface respectively.

**Comment by clebertsuconic** [comment]
That's has always been a server destination.  There is a code for that kind of thing in the thing in the converter.  


You are adding a brand new test thought.  How could it be failing on old changes?


I will have to look into this tomorrow my morning.

**Comment by gaohoward** [comment]
The changes was commited at Mon Apr 25 (Andy), after I sent my PR. I didn't run my test after rebase, which I should have. :)

**Comment by gaohoward** [comment]
ok, I'll leave that to you. @clebertsuconic

**Comment by clebertsuconic** [comment]
>> The changes was commited at Mon Apr 25 (Andy), after I sent my PR. I didn't run my test after rebase, which I should have. :)

No big deal.. I didn't have the whole picture... thanks for this commit, nice one!

**Pull request closed without merging**